Add enable_gpu_isolation flag

klueska opened the pull request:
This patch creates a new config.yaml parameter called `enable_gpu_isolation` that enables (cluster-wide) GPU isolation from mesos. This means that *all* machines in your cluster must have GPUs installed on them in order to enable this feature. Otherwise starting the mesos-agent on non-GPU capable machines will fail.

For now, this support is considered experimental, so requiring all machines in a cluster to have GPUs is acceptable for testing out the feature. In future releases we will proide support for having a mixed cluster of GPU and non GPU machines, where only the GPU-capable machines have this feature enabled.

cmaloney commented:
Will need a manual test + confirmation the advanced aws from dgc.sh work, as well as a manual install on a VPC cluster

cmaloney commented on gen/calc.py:
```diff
@@ -326,6 +326,13 @@ def calculate_config_yaml(user_arguments):
 
 
+def calculate_mesos_isolation(enable_gpu_isolation):
+    isolators = 'cgroups/cpu,cgroups/mem,disk/du,network/cni,filesystem/linux,docker/runtime,docker/volume',
```
I accidentally left a trailing comma here which caused this to be a tuple rather than a string, if you remove the comma the errors should go away

cmaloney commented:
Holding until hand tests are complete

klueska commented:
I completed the manual tests. A journal of my test procedure is below:

```
I followed some subset of the information at the following links to bring up
my DC/OS cluster for testing with GPUs.

1) Generating Custom AWS CF Templates
   https://dcos.io/docs/1.8/administration/installing/cloud/aws/advanced/aws-custom/

2) Quick Start Guide
   ttps://dcos.io/docs/1.8/administration/installing/cloud/aws/advanced/quickstart/

The steps I followed are below:

-- I started by follow instructions from:
   https://dcos.io/docs/1.8/administration/installing/cloud/aws/advanced/aws-custom/

-- I created new S3 bucket called 'klueska-advanced-aws-gpu'

-- I downloaded the relevant dcos_generate_config.sh
   $ wget https://downloads.dcos.io/dcos/testing/pull/766/dcos_generate_config.sh

-- I made a genconf directory and added a config.yaml to it
   ---
   enable_gpu_isolation: true
   aws_template_storage_bucket: klueska-advanced-aws-gpu
   aws_template_storage_bucket_path: templates/dcos
   aws_template_storage_region_name: us-west-2
   aws_template_upload: true
   aws_template_storage_access_key_id: <my-access-key-id>
   aws_template_storage_secret_access_key: <my-secret-access_key>

-- I ran the following command
   $ sudo bash dcos_generate_config.sh --aws-cloudformation

-- I went to my S3 bucket and navigated to:
   -- klueska-advanced-aws-gpu/templates/

-- I right clicked on 'dcos' and selected 'Make public'

-- I navigated to the following and clicked on 'Properties' for 'el7-zen-1.json':
   klueska-advanced-aws-gpu/templates/dcos/config_id/60c2916fc5d6eac129bf72a692392e2e9f9d3a1c/cloudformation

-- I downloaded 'el7-zen-1.json'

-- I modified all references of TemplateURL to manually fix the known bug DCOS-10279
   (4 instances of it)

-- I re-uploaded 'el7-zen-1.json' to its original location in my S3 bucket

-- I copied the fillowing URL to this template:
   https://s3-us-west-2.amazonaws.com/klueska-advanced-aws-gpu/templates/dcos/config_id/60c2916fc5d6eac129bf72a692392e2e9f9d3a1c/cloudformation/el7-zen-1.js

-- I went to CloudFormation -> Create Stack (per the instructions)

-- I pasted my custom template URL on the "Select Template" page and clicked next

-- I jumped to the AWS Advanced "Quick Start Guide"
   https://dcos.io/docs/1.8/administration/installing/cloud/aws/advanced/quickstart/

   Note: There were no instructions tellimg me to do this, I just knew to do it.

-- I created a zen.sh file and copied the script from the instructions to it

-- I ran the zen script as:
   $ bash ./zen.sh klueska-gpu-test

-- I filled in the fields on the Cloud Formation "Specify Details" page
   -- I had to refresh this page before filling in the fields
      (Otherwise I wasn't able to autocomplete on the subnet/vpc names)

   Stack Name: DCOS-GPU-test
   AdminLocation: 0.0.0.0/0
   CustomAmi: ami-9b5d97fb (this is our standard CentOS image with Nvidia GPU tools installed)
   InternetGateway: igw-07823363 (custom output from zen)
   KeyName: <my-key-name>
   MasterInstanceType: m3.xlarge
   OAuthEnabled: true
   PrivateAgentInstanceCount: 1
   PrivateAgentInstanceType: g2.2xlarge
   PrivateSubnet: subnet-9cf0acea (custom output from zen)
   PublicAgentInstanceCount: 1
   PublicAgentInstanceType: g2.2xlarge
   PublicSubnet: subnet-98f0acee (custom output from zen)
   Vpc: vpc-1a5f437e (custom output from zen)

-- Under "Options", I added some tags to make sure my cluster stayed up for a while:

   owner = klueska
   expiration = 4h

-- I started creation of the stack!

-- Once up, I navigated to my cluster per the instructions
  (It took 45 minutes to come up!)

-- I deployed 1 simple GPU instance and 1 docker based GPU instance with marathon
   (I wanted to test both public and private slave instances)

   Simple:

   {
     "id": "simple-gpu-test",
     "acceptedResourceRoles":["slave_public", "*"],
     "cmd": "while [ true ] ; do nvidia-smi; sleep 5; done",
     "cpus": 1,
     "mem": 128,
     "disk": 0,
     "gpus": 1,
     "instances": 1
   }

   Docker-based:

   {
     "id": "docker-gpu-test",
     "acceptedResourceRoles":["slave_public", "*"],
     "cmd": "while [ true ] ; do nvidia-smi; sleep 5; done",
     "cpus": 1,
     "mem": 128,
     "disk": 0,
     "gpus": 1,
     "instances": 1,
     "container": {
       "type": "MESOS",
       "docker": {
         "network": "HOST",
         "image": "nvidia/cuda"
       }
     }
   }

-- The simple instance came up and started running immediately

-- The docker instance failed with the error:
   mesos-containerizer: error while loading shared libraries: libssl.so.1.0.0: cannot open shared object file: No such file or directory

-- I ssh'd into both the private and the public agent and ran:
   $ sudo ldconfig /opt/mesosphere/lib

-- Then I redeployed the GPU docker test and it worked.
```

klueska commented:
Is there any more I need to do for this?

cmaloney commented:
nope, it's mainly just waiting for 1.8.5 to ship so that master is headed towards 1.9

ryankennedyio commented:
@klueska is there any chance you could make that CentOS w/ Nvidia AMI public? I have followed along those steps & all seems fine, but no nodes are able to be detected from DC/OS UI after launch. That was the only thing differed from your steps. I used ami-70017967 as a base, installed CUDA correctly and snapshotted the image. Unsure what else it could be.

cmaloney commented:
@ryankennedyio at this point in time every agent in the cluster must be GPU-enabled. One of the constraints that make this experimental.

klueska commented:
I take that back. It looks like I do have access to it from my personal
account. Do you see it when you try to launch a new EC2 instance and,
search under "Community AMIs" for it?

On Tue, Oct 25, 2016 at 8:24 AM Kevin Klues <klueska@gmail.com> wrote:

> I didn't make the image, but I'll see if I have permission to make it
> public. I asked the person who made it to make it public (and they said
> they did), but it looks like it actually isn't (I just tried finding it in
> the AWS marketplace from my personal account and it's not there).
>
> On Mon, Oct 24, 2016 at 11:26 PM Ryan Kennedy <notifications@github.com>
> wrote:
>
> @klueska <https://github.com/klueska> is there any chance you could make
> that CentOS w/ Nvidia AMI public? I have followed along those steps & all
> seems fine, but no nodes are able to be detected from DC/OS UI after
> launch. That was the only thing differed from your steps. I used
> ami-70017967 as a base, installed CUDA correctly and snapshotted the image.
> Unsure what else it could be.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dcos/dcos/pull/766#issuecomment-255947802>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAF4o6mNGtEMr3YduM-ydXXZaPiC2d57ks5q3aEGgaJpZM4KEeJX>
> .
>
>

ryankennedyio commented:
Hi @klueska, weird. I get the following screenshot. 
I just tried to re-launch the stack with that AMI specified, but it fails at the MasterStack step.

![screen shot 2016-10-26 at 4 59 50 am](https://cloud.githubusercontent.com/assets/11158006/19699020/1f998f1e-9b39-11e6-93f6-153118a6963c.png)

klueska commented:
I didn't make the image, but I'll see if I have permission to make it
public. I asked the person who made it to make it public (and they said
they did), but it looks like it actually isn't (I just tried finding it in
the AWS marketplace from my personal account and it's not there).

On Mon, Oct 24, 2016 at 11:26 PM Ryan Kennedy <notifications@github.com>
wrote:

> @klueska <https://github.com/klueska> is there any chance you could make
> that CentOS w/ Nvidia AMI public? I have followed along those steps & all
> seems fine, but no nodes are able to be detected from DC/OS UI after
> launch. That was the only thing differed from your steps. I used
> ami-70017967 as a base, installed CUDA correctly and snapshotted the image.
> Unsure what else it could be.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dcos/dcos/pull/766#issuecomment-255947802>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAF4o6mNGtEMr3YduM-ydXXZaPiC2d57ks5q3aEGgaJpZM4KEeJX>
> .
>

ryankennedyio commented:
No problem @klueska, appreciate it a lot.

klueska commented:
Looks like the AMI is only available in US West (Oregon). Let me see what
it will take to copy it to other regions. What region are you in?

On Tue, Oct 25, 2016 at 12:03 PM Ryan Kennedy <notifications@github.com>
wrote:

> No problem @klueska <https://github.com/klueska>, appreciate it a lot.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dcos/dcos/pull/766#issuecomment-256143533>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAF4o0D9Xj6AuREN-PBGMzG_IePuqXYfks5q3lJ9gaJpZM4KEeJX>
> .
>

klueska commented:
It looks like you are based in Australia. I copied the AMI to the Asia Pacific (Sydney) region. The copy has an id of ami-37b28f54

Let me know if this works for you.

ryankennedyio commented:
Sorry @klueska we are set up at east-us-1 -- possible to move it there? Otherwise need to wait a bit longer for EC2 limit requests on G2 or P2 instances. Have access to it just fine now though.

klueska commented:
New image added to east-us-1: ami-e10e50f6

ryankennedyio commented:
Works beautifully. Thanks so much @klueska. Very excited for the 1.9 release!

yuefengz commented:
What do `Nvidia GPU libraries` refer to? The nvidia driver and CUDA and CuDNN?

klueska commented:
They refer to libnvidia-ml
https://developer.nvidia.com/nvidia-management-library-nvml

klueska commented:
Instructions on how to install it (and verify the installation) are here:
https://github.com/apache/mesos/blob/master/docs/gpu-support.md#external-dependencies

yuefengz commented:
@klueska Thanks! Looks like driver has the NVML library.
